### DISCUSSION TITLE
typical build environment configuration

### DISCUSSION THREAD

**Discussion by digiexchris** [UncleRus/esp-idf-lib#213] [Q&A] [answered]
Hi! I'm relatively new to esp-idf, and I'm wondering how people typically structure they're projects to include something like esp-idf-lib where multiple components have interconnected dependencies, such as the i2c version of the hh44780 and i2cdev. 

I have a CMakeList.txt that contains
set(COMPONENT_ADD_INCLUDEDIRS "C:/Users/chris/Desktop/esp-idf/esp-idf-lib/components/.") 

Which seems to satisfy my include for the hh44780.h (which I have an explicit include for) but hh44780 depends on i2cdev.h which it's unable to find. I have the esp-idf-lib in a different place than my esp-idf components so I can easily upgrade them individually. I was hoping to include them separately. I notice a lot of the examples don't need to have specific includedirs in their CMakeLists.txt so I'm wondering how people typically do this?

**Comment by UncleRus** [comment, marked as answer]
ESP-IDF project contains at least two CMakeList.txt: one for the project and one for the `main` component:
- `YourProject/main/CMakeList.txt`
- `YourProject/CMakeList.txt`

CMakeLists.txt in `main` usually contains only the sources of your project:
```
idf_component_register(
        INCLUDE_DIRS .
        SRCS main.c settings.c ....
        EMBED_TXTFILES web/jquery.js web/styles.css ...
        EMBED_FILES gifs/heart.gif gifs/heart2.gif ...
)
```
And your project CMakeList.txt can contains links to other libaries, for example:
```
cmake_minimum_required(VERSION 3.5)

set(EXTRA_COMPONENT_DIRS C:/Users/chris/Desktop/esp-idf/esp-idf-lib/components/)

include($ENV{IDF_PATH}/tools/cmake/project.cmake)
project(MyProject)
```

You can see an example here: https://github.com/UncleRus/EvLamp

**Reply by digiexchris** [reply]
thank you! that solved all of my problems. That gave me the two pieces of info that I was missing! 
1: the project level CMakeList.txt gets the other libraries
2: EXTRA_COMPONENT_DIRS rather than INCLUDE_DIRECTORIES

**Reply by UncleRus** [reply]
You are welcome!
ESP-IDF build system described here: 
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system.html
